### PR TITLE
Splitting Links into libraries

### PR DESCRIPTION
**Pull request by frank-emrich**
This PR realizes something I've mentioned a few times: Splitting Links into individual dune libraries beyond just core and bin.

In particular, the layout introduced by this PR is as follows:

# core
A subset of the old `core` library, essentially the basic shared functionality.
In particular, the types are in here
Dependencies: none

# frontend
Contains the sugar AST, the frontend typechecker, and the sugar transformations
Dependencies: core

# backend
Contains the IR and the IR typechecker
Dependencies: core

# middle-end
This contains the translations between frontend and backend representations. In particular, `SugarToIr` lives here.
Dependencies: core, frontend, backend

# runtime
Contains the evaluator, `Value` and friends, database stuff, query handling and everything web-related
Dependencies: core, frontend, and backend. The dependency on frontend is due to the fact that some modules in here rely on the frontend parser for types. 

# bin
As before, this contains the REPL, the links executable and the driver 
Dependencies: all other libraries

Not mentioned above: The library `lens` remains as it was and all other libraries use it

### CODE REVIEW

**Comment by frank-emrich** [comment]
The only change that was necessary beyond just moving files around and opening the right libraries is the following:

Currently, the `Lib` module provides both typing information about the library functions as well as their implementation. Due to the latter, it must know about `Value`, and therefore sits in the `runtime` library.

However, some earlier stages of Links need access to the typing information, namely `Sugartoir` and `IrCheck`. I've therefore functorized those two modules over a small signature, called `LIB_TYPING_INFO`, that provides some functions about the typing of `Lib` functions.

**Comment by SimonJF** [comment]
Planning to have a detailed look at this tomorrow morning. Thanks so much -- this looks like an excellent step, and would also in my opinion close #629. I'm glad that the functorisation step removed the need for the hack you had.

The CI seems to be broken due to a unit test dependency. Could you take a look?

I think before merging this, we should probably discuss it at a Links developer meeting.

**Comment by frank-emrich** [comment]
> The CI seems to be broken due to a unit test dependency. Could you take a look?

This was unrelated; a new release of ounit broke things. I address this is in #763

**Comment by SimonJF** [comment]
@frank-emrich -- would you mind sorting the merge conflicts here so we can get CI going again? Thanks!

**Comment by frank-emrich** [comment]
done!

**Comment by SimonJF** [comment]
Brilliant, thanks!

File: backend/dune [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
```
[the remark this answers is not in the document]
Comment (frank-emrich) [reply backend/dune]: I've usually added some empty lines between the `open`s for the Links libraries and the `open`s for the actual modules. But there were also a lot of random new lines, I've deleted those

**Comment by frank-emrich** [comment]
> Quick question though: why is `instantiate.ml` here? Does that not belong in the frontend?

That's because it both contains general type instantiation logic (which is also used) by the IR type-checker, but also some logic more specific to type inference. 

> I wonder whether it may be worth separating out some of the query compilation files from runtime, since one might argue that they are a substantial enough part of runtime to warrant their own library.

I haven't tried it yet and don't know if it's possible to factor the query stuff out of `runtime` easily.
It seems like `Query` and `EvalQuery` depend on `Value`, and `Evalir` depends on `EvalQuery`. I think this is also a matter of taste, where some people may prefer keeping the number of overall libraries low.


> Could you define this in Types instead?

Done (The github UI is broken, I can't respond to the comment atm)

**Comment by slindley** [comment]
Thanks @SimonJF. You're quite right. This is a significant enough change that we should discuss it at a developer meeting before it is merged. I'll have a look at it once I get back from holiday.

**Comment by frank-emrich** [comment]
That shouldn't be a problem. Since there are only minor actual changes to the files besides them being moved around, keeping this up-to-date if we don't merge immediately is easy.

**Comment by jamescheney** [comment]
I'd like the query stuff to eventually be its own library (or eventually done in the frontend by reduction to other features).

I'd also like #624 to be merged first before this PR though.  I think #624 is only waiting on some performance evaluation taking place to understand how performance changes, if at all.

**Comment by jstolarek** [comment]
Also:

> I'd also like #624 to be merged first before this PR though.

#624 seems orthogonal?

**Comment by frank-emrich** [comment]
> 1. As stated yesterday, I would prefer to have open `Link_core.foo` instead of ...

The problem is that in general, those are not equivalent. If we don't do `open Links_core`, you would have to prefix every module from the `core` library with `Links_core`. This is a matter of taste, but to me it seems more convenient to start each source file with `open` statements for all the libraries it uses rather than using fully qualified module names throughout the rest of the file. I don't expect us to ever have more than a handful of libraries and if two libraries contained modules with the same name we should refer to those with fully qualified names.

> 2. I really don't like having two empty lines between open statements. If (1) above is done this problem will go away.

The idea was to visually separate the part where we open libraries and where we open actual modules. Again a matter of taste I guess.

> 3. We have `frontend` and `backend`, but then you say `middle_end` and not `middleend`.

True, that's inconsistent. I used `middleend` first, but it kind of hurt my eyes. I'm fine with going back to `middleend` if people prefer that.

> 4.It would be nicer to have `Links.Core` or `LinksCore` instead of `Links_core`. Do we have any control over that?

Unfortunately not, this is what dune dictates.


> #624 seems orthogonal?

I guess James was worried about merge conflicts, which I think are not such a big problem because git detects that the files are just moved. In any case, I'm happy to wait until that PR is merged.

**Comment by jamescheney** [comment]
> I guess James was worried about merge conflicts,

Yes

**Comment by slindley** [comment]
The idea is fine. I'm not at all happy with the names though, as they're very misleading. First "middle-end" is certainly not a thing! Second, some of the other names are inaccurate.

I have no problem with `core` and `bin`.

The `frontend` library at least contains the frontend, but it goes a bit beyond that. It is really a combination of the frontend and a collection of desugaring transformations. OK, I guess this name is fine.

The `backend` library is **not** a backend at all. It should be called `ir`.

The `middle-end` library is quite small. It translates the frontend representation (which tends to be called `sugar` throughout the code base) into the IR. It could be called `frontendtoir`. I guess the reason for it to exist is to allow `ir` to be completely decoupled from `frontend`.

The `runtime` library contains various backends and mixes these up with various parts of the runtime because Links is still interpreted on the server. The `bin` library also includes important parts of the runtime. The `runtime` library should be called `backend`. Ultimately, it should be split up into several smaller libraries including the query backend.

**Comment by jamescheney** [comment]
This is out of date and delayed to after #637  and we just discussed redoing and either closing/reopening or making a new PR.

**Comment by frank-emrich** [comment]
Closing until I have a chance to redo it on top of current Links

**Pull request closed without merging**